cscs-checks/microbenchmarks/gpu/pointer_chase/src/linked_list.hpp

```diff
@@ -0,0 +1,356 @@
+{
+  /* List random initializer
+   * - buffer: where the list is to be placed.
+   * - indices: array containing the node ordering indices as offsets in
+   *     the buffer.
+   */
+
+  // Set the head
+  Node * prev = buffer + indices[0];
+
+  // Init the rest of the list
+  for (int n = 1; n < num_nodes; n++)
+  {
+    Node * temp = buffer + indices[n];
+    prev->next = temp;
+    prev = temp;
+  }
+  prev->next = buffer + indices[0];
+
+}
+
+__global__ void simple_traverse(Node * __restrict__ buffer, uint32_t head_index)
+{
+  /* Simple list traverse - no timing is done here
+   * - buffer: where the list is
+   * - head_index: index in the buffer where the head of the list is
+   */
+
+  Node * head = &(buffer[head_index]);
+  Node * ptr = head;
+  while(ptr->next != head)
+  {
+    ptr = ptr->next;
+  }
+
+  // Silly dep. to tell the compiler not to throw away this kernel.
+  if (ptr->next->next == head)
+  {
+    printf("The impossible just happened\n");
+  }
+
+}
+
+
+#ifdef VOLATILE
+# define __VOLATILE__ volatile
+#else
+# define __VOLATILE__
+#endif
+
+/*
+ * Timed list traversal. This implementation is recursive (because it's less code) so you have to
+ * watch out to not exceed the recursion limits. The functions are force-inlined, so the PTX code
+ * looks identical as if you were to unwrap the recursion manually.
+ *
+ * Depending on the compiler flags used, the timing can either measure each node jump, or the entire
+ * list traversal as a whole.
+ */
+__device__ __forceinline__ void next_node( __VOLATILE__ Node ** ptr, uint32_t * timer, Node ** ptrs, int & jump)
+{
+  /*
+   * Recursive function to traverse the list.
+   * - ptr: Pointer of a pointer to a node in the linked list.
+   * - timer: Array to store the timings of each individual node jump.
+   *   Only used if this option is activated (-DTIME_EACH_STEP)
+   * - ptrs: Just used to have a data dependency to block ILP.
+   * - jump: Int to keep track of the number of jumps
+   */
+
+# ifdef TIME_EACH_STEP
+  XClocks64 clocks;
+  clocks.start();
+# endif
+  (*ptr) = (*ptr)->next;
+# ifdef TIME_EACH_STEP
+  *(ptrs+jump) = (*ptr);  // Data dep. to prevent ILP.
+  *(timer+jump) = clocks.end();    // Time the jump
+# endif
+
+  jump++;
+}
+
+
+__global__ void timed_list_traversal(Node * __restrict__ buffer, uint32_t head_index, uint32_t * timer)
+{
+  /* Timed List traversal - we make a singly-linked list circular just to have a data dep. and
+   * cover from compiler optimisations.
+   */
+
+  // These are used to prevent ILP when timing each jump.
+  __shared__ uint32_t s_timer[JUMPS];
+  __shared__ Node * ptrs[JUMPS];
+
+  // Create a pointer to iterate through the list
+  __VOLATILE__ Node * ptr = &(buffer[head_index]);
+
+  // Node jump counter
+  int jump = 0;
+
+#ifndef TIME_EACH_STEP
+  // start timer
+  XClocks64 clocks;
+  clocks.start();
+#endif
+
+  // Traverse the list
+  REPEAT_JUMPS(next_node(&ptr, s_timer, ptrs, jump);)
```

Comment:
I would really strongly discourage you from using macros for essential code-logic. The list traversal is THE core essential part of the entire benchmark and this part is implemented here with the macro `REPEAT_JUMPS` that's not even defined in the same file. It's perfectly adequate to use a plain loop here:
```suggestion
 while(num_jumps--)
 {
     buffer = buffer->next;
 }
```
where `num_jumps` is the number of jumps that I've added as a command line parameter in my tests.

The same goes for `JUMPS`. Together with the number of nodes, the number of jumps is the key parameter of the benchmark and there is simply no way around making it dynamically controllable for achieving accurate results.

In general, if you really must use a macro, my advice is to only define/reference them within the same file.